pageserver/src/config.rs

```diff
@@ -1359,6 +1363,58 @@ trace_read_requests = {trace_read_requests}"#,
         Ok(())
     }
 
+    #[test]
+    fn parse_incorrect_tenant_config_with_change() -> anyhow::Result<()> {
+        let tempdir = tempdir()?;
+        let (workdir, pg_distrib_dir) = prepare_fs(&tempdir)?;
+
+        let config_string = format!(
+            r#"
+        {ALL_BASE_VALUES_TOML}
+        pg_distrib_dir='{}'
+        [tenant_config]
+        checkpoint_distance = -1 # supposed to be an u64
+    "#,
+            pg_distrib_dir.display()
+        );
+
+        let toml = config_string.parse()?;
+
+        let error = PageServerConf::parse_and_validate(&toml, &workdir).unwrap_err();
+
+        let expected_error_str = "checkpoint_distance: invalid value: integer `-1`, expected u64\nin `checkpoint_distance`\n";
```

Comment:
Ok, this was just very surprising that `toml_edit::parser::errors::TomlError` has so verbose display impl. Of course, it can probably be worked around to be single line, probably. Luckily we can just access the message ("invalid value: integer `-1`, expected u64") via `TomlError::message(&self)`.

I was also confused because of course we are able to use serde with TenantConf because that is how the json is parsed. Phew.

So doesn't these two test cases show that we could perhaps with very simple changes to drop the manual TenantConf parsing code?

Reply:
Yes, we can replace it
There is one more place where manual parsing is going on `TryFrom<&'_ models::TenantConfig> for TenantConfOpt`, i tried replacing it with `serde_path_to_error` deserializer in  api handler but it doesn't detect path for flattened fields with incorrect values